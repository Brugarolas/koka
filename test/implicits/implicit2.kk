// with expression syntax
implicit val width: int
implicit val height: int

val res1 = with { val height = 5 } in { height + height }

val res2 = with val height = 5 in height + height

val res3 = with height = 5 in height + height

val res4 = with height = 5 in
           with width = 7 in height + width

// the optimizer inlines `1 + 3` into `height() { resume(1 + 3) }` which
// is problematic when accessing the state.
val res5 = with (s = 7) { val height = 1 + 3 } in height * height

fun res6() {
  (with (s = 7) { val height = { println("foo"); 1 + 3 } } in height * height)
}

effect counter {
  fun tick(): ()
  fun get(): int
}

val res7 = with (s = 0) {
  effect fun tick() { resume((), s + 1) }
  effect fun get() { resume(s, s) }
} in {
  tick()
  tick()
  tick()
  get()
}

// with statement syntax
val res8 = {
  with { val width = 5 }
  width + width
}

val res9 = {
  with { val width = 5 }
  with { val height = 9 }
  width * height
}

val res10 = {
  with width = 5
  with height = 9
  width * height
}

effect amb { fun flip(): bool }

val returnClause = {
  with height = 4
  with flip() -> (resume(True) : list<_>) + resume(False)
  with return a -> [a]

  if (flip()) 3 else height
}

fun evaluationOrderStmt() {
  with height = { println("height"); 4 }
  with width  = { println("width"); 7 }
  height + width
}

fun evaluationOrderExpr() {
  (with height = { println("height"); 4 } in
   with width  = { println("width"); 7 } in
   height + width)
}

fun main() {
  res6()
  evaluationOrderStmt().println
  evaluationOrderExpr().println
}
