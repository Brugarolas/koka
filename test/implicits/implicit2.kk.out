[01;30moperator branch (evaluationOrderExpr) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (evaluationOrderExpr) .val-width: resume tail
[00;39;49m[00m[01;30moperator branch (evaluationOrderStmt) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (evaluationOrderStmt) .val-width: resume tail
[00;39;49m[00m[01;30moperator branch (res6) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (res7) tick: resume tail
[00;39;49m[00m[01;30moperator branch (res7) get: resume tail
[00;39;49m[00m[01;30moperator branch (res1) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (res10) .val-width: resume tail
[00;39;49m[00m[01;30moperator branch (res10) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (res2) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (res3) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (res4) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (res4) .val-width: resume tail
[00;39;49m[00m[01;30moperator branch (res8) .val-width: resume tail
[00;39;49m[00m[01;30moperator branch (res9) .val-width: resume tail
[00;39;49m[00m[01;30moperator branch (res9) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (returnClause) .val-height: resume tail
[00;39;49m[00m[01;30moperator branch (returnClause) flip: resume scoped
[00;39;49m[00mfoo
16
3
height
width
11
height
width
11
.Op-flip : .op-flip
.Op-get : .op-get
.Op-tick : .op-tick
.Op-val-height: .op-val-height
.Op-val-width: .op-val-width
.Ops-flip : forall<a> (flip : .op-flip) -> .ops-amb<bool>
.Ops-get : forall<a> (get : .op-get) -> .ops-counter<int>
.Ops-tick : forall<a> (tick : .op-tick) -> .ops-counter<()>
.Ops-val-height: forall<a> (.op-val-height) -> .ops-height<int>
.Ops-val-width: forall<a> (.op-val-width) -> .ops-width<int>
.tag-amb : string
.tag-counter: string
.tag-height : string
.tag-width : string
.val-height : () -> height int
.val-width : () -> width int
evaluationOrderExpr: () -> console int
evaluationOrderStmt: () -> console int
flip : () -> amb bool
get : () -> counter int
height : value<int>
main : () -> console ()
res1 : int
res10 : int
res2 : int
res3 : int
res4 : int
res6 : () -> console int
res7 : () -> int
res8 : int
res9 : int
returnClause: list<int>
tick : () -> counter ()
width : value<int>